asmcomp/deadcode.ml

```diff
@@ -72,65 +68,40 @@ let rec deadcode i =
       { i = {i with desc = Iswitch(index, cases'); next = s.i};
         regs = Reg.add_set_array i.live arg;
         exits = Array.fold_left
-                  (fun acc c -> IntSet.union acc c.exits) s.exits dc;
+                  (fun acc c -> Int.Set.union acc c.exits) s.exits dc;
       }
   | Icatch(rec_flag, handlers, body) ->
     let body' = deadcode body in
     let s = deadcode i.next in
-    let (handlers', h_exits) =
-      List.split
-        (List.map (fun (nfail, handler) ->
-           let handler' = deadcode handler in
-           ((nfail, handler'.i), handler'.exits))
-           handlers)
+    let handlers' = Int.Map.map deadcode (Int.Map.of_list handlers) in
+    let rec add_live exit (live_exits, used_handlers) =
+      if Int.Set.mem exit live_exits then
+        (live_exits, used_handlers)
+      else
+        let live_exits = Int.Set.add exit live_exits in
+        match Int.Map.find_opt exit handlers' with
+        | None -> (live_exits, used_handlers)
+        | Some handler ->
+            Int.Set.fold add_live handler.exits
+              (live_exits, (exit, handler.i) :: used_handlers)
     in
-    (* Remove unused handlers.
-       We deal with three disjoint sets of "nfail" indexes:
-       dead handler indexes = B \setminus A
-       used handler indexes = A \intersect B
-       live exit indexes = A \setminus B
-       where
-       B = indexes of handlers in this Icatch,
-       A = indexes used in Iexit instructions that are
-       in scope of the handlers of this Icatch (scope depends on rec_flag).
-    *)
-    let h_exits = List.fold_left (fun acc e -> IntSet.union acc e)
-                    IntSet.empty h_exits in
-    let all_exits = IntSet.union body'.exits h_exits in
-    let exits_in_scope =
-      match rec_flag with
-      | Cmm.Recursive -> all_exits
-        (* Indexes mentioned in Iexit instructions in body and handlers. *)
-      | Cmm.Nonrecursive -> body'.exits
-        (* Icatch body is allowed to Iexit to it's own hanlders
-           or handlers of an enclosing Icatch that are not shadowed
-           by the current Icatch.
-           A handler cannot Iexit to itself or other handlers in this Icatch.
-           Thus, all Iexits in handlers refer to an enclosing Icatch. *)
+    let live_exits, used_handlers =
+      Int.Set.fold add_live body'.exits (s.exits, [])
```

Comment:
I thought that the handlers of i.next are not in scope in this Icatch. If so, s.exits should be replaced with body'.exists here, and keep the "union" as before, unless nfail numbers are unique across the entire function, and not just in scope (which is what I assumed before).

Reply:
Yes that's the case, but since I also assumed here that the input is correct, I could just do the union by using s.exits here.
That probably need some comments.

Reply:
great, thanks! I'll add a comment.